File: node/test/replayExpansion.test.ts
import { describe, it, expect } from "vitest";
import { pool } from "../src/db/pool.js";
import { runReplayExpansionCycle } from "../src/worker/replayExpansion.js";
import { createTenant, createEndpoint, createDelivery } from "./fixtures.js";

async function createReplay(
  endpointId: string,
  rangeStart: Date,
  rangeEnd: Date,
  idempotencyKey = `replay-fixture-${crypto.randomUUID()}`,
): Promise<string> {
  const { rows } = await pool.query<{ id: string }>(
    `INSERT INTO replays (endpoint_id, idempotency_key, range_start, range_end) VALUES ($1, $2, $3, $4) RETURNING id`,
    [endpointId, idempotencyKey, rangeStart, rangeEnd],
  );
  return rows[0]!.id;
}

const windowStart = new Date(Date.now() - 60_000);
const windowEnd = new Date(Date.now() + 60_000);
const beforeWindow = new Date(Date.now() - 120_000);

describe("runReplayExpansionCycle", () => {
  it("creates a fresh delivery row per terminal original in the window and flips the replay to expanded", async () => {
    const { id: tenantId } = await createTenant();
    const endpoint = await createEndpoint(tenantId, ["order.created"]);
    const original = await createDelivery(tenantId, endpoint.id, { state: "succeeded", createdAt: windowStart });
    const replayId = await createReplay(endpoint.id, windowStart, windowEnd);

    const didWork = await runReplayExpansionCycle(pool);

    expect(didWork).toBe(true);

    const { rows: replayRows } = await pool.query<{ status: string }>("SELECT status FROM replays WHERE id = $1", [replayId]);
    expect(replayRows[0]!.status).toBe("expanded");

    const { rows: deliveryRows } = await pool.query<{ id: string; event_id: string; state: string }>(
      "SELECT id, event_id, state FROM deliveries WHERE endpoint_id = $1 AND id != $2",
      [endpoint.id, original.id],
    );
    expect(deliveryRows).toHaveLength(1);
    expect(deliveryRows[0]!.event_id).toBe(original.eventId);
    expect(deliveryRows[0]!.state).toBe("pending");
    expect(deliveryRows[0]!.id).not.toBe(original.id); // fresh delivery_id, per R-19/R-20
  });

  it("excludes originals outside the replay window", async () => {
    const { id: tenantId } = await createTenant();
    const endpoint = await createEndpoint(tenantId, ["order.created"]);
    await createDelivery(tenantId, endpoint.id, { state: "succeeded", createdAt: beforeWindow });
    await createReplay(endpoint.id, windowStart, windowEnd);

    await runReplayExpansionCycle(pool);

    const { rows } = await pool.query<{ count: string }>(
      "SELECT count(*) FROM deliveries WHERE endpoint_id = $1 AND state != 'succeeded'",
      [endpoint.id],
    );
    expect(Number(rows[0]!.count)).toBe(0);
  });

  it("excludes still-pending and in_flight originals, since they'll be attempted on their own schedule regardless", async () => {
    const { id: tenantId } = await createTenant();
    const endpoint = await createEndpoint(tenantId, ["order.created"]);
    await createDelivery(tenantId, endpoint.id, { state: "pending", createdAt: windowStart });
    await createDelivery(tenantId, endpoint.id, { state: "in_flight", createdAt: windowStart });
    await createReplay(endpoint.id, windowStart, windowEnd);

    await runReplayExpansionCycle(pool);

    const { rows } = await pool.query<{ count: string }>(
      "SELECT count(*) FROM deliveries WHERE endpoint_id = $1",
      [endpoint.id],
    );
    expect(Number(rows[0]!.count)).toBe(2); // only the two originals, no replayed rows added
  });

  it("inserts replayed deliveries in the original chronological order (by seq)", async () => {
    const { id: tenantId } = await createTenant();
    const endpoint = await createEndpoint(tenantId, ["order.created"]);
    const first = await createDelivery(tenantId, endpoint.id, { state: "succeeded", createdAt: windowStart });
    const second = await createDelivery(tenantId, endpoint.id, { state: "failed", createdAt: windowStart });
    await createReplay(endpoint.id, windowStart, windowEnd);

    await runReplayExpansionCycle(pool);

    const { rows } = await pool.query<{ event_id: string }>(
      "SELECT event_id FROM deliveries WHERE endpoint_id = $1 AND state = 'pending' ORDER BY seq",
      [endpoint.id],
    );
    expect(rows.map((r) => r.event_id)).toEqual([first.eventId, second.eventId]);
  });

  it("returns false when there is nothing pending to expand", async () => {
    const didWork = await runReplayExpansionCycle(pool);
    expect(didWork).toBe(false);
  });
});

File: go/internal/worker/replay_expansion_test.go
package worker_test

import (
	"context"
	"testing"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/stretchr/testify/require"

	"webhooks-go/internal/testsupport"
	"webhooks-go/internal/worker"
)

var (
	replayWindowStart  = time.Now().Add(-60 * time.Second)
	replayWindowEnd    = time.Now().Add(60 * time.Second)
	replayBeforeWindow = time.Now().Add(-120 * time.Second)
)

func insertReplay(t *testing.T, pool *pgxpool.Pool, endpointID string, rangeStart, rangeEnd time.Time) string {
	t.Helper()
	var id string
	err := pool.QueryRow(context.Background(),
		`INSERT INTO replays (endpoint_id, idempotency_key, range_start, range_end) VALUES ($1, $2, $3, $4) RETURNING id`,
		endpointID, "replay-fixture-"+uuid.NewString(), rangeStart, rangeEnd,
	).Scan(&id)
	require.NoError(t, err)
	return id
}

func TestRunReplayExpansionCycleCreatesFreshDeliveryPerTerminalOriginal(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ctx := context.Background()
	tenantID, _ := testsupport.CreateTenant(t, pool)
	endpointID := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})
	originalID, originalEventID := testsupport.CreateDelivery(t, pool, tenantID, endpointID, testsupport.DeliveryOptions{
		State: "succeeded", CreatedAt: &replayWindowStart,
	})
	replayID := insertReplay(t, pool, endpointID, replayWindowStart, replayWindowEnd)

	did, err := worker.RunReplayExpansionCycle(ctx, pool)
	require.NoError(t, err)
	require.True(t, did)

	var status string
	require.NoError(t, pool.QueryRow(ctx, "SELECT status FROM replays WHERE id = $1", replayID).Scan(&status))
	require.Equal(t, "expanded", status)

	rows, err := pool.Query(ctx, "SELECT id, event_id, state FROM deliveries WHERE endpoint_id = $1 AND id != $2", endpointID, originalID)
	require.NoError(t, err)
	var replayed []struct {
		ID      string
		EventID string
		State   string
	}
	for rows.Next() {
		var r struct {
			ID      string
			EventID string
			State   string
		}
		require.NoError(t, rows.Scan(&r.ID, &r.EventID, &r.State))
		replayed = append(replayed, r)
	}
	require.Len(t, replayed, 1)
	require.Equal(t, originalEventID, replayed[0].EventID)
	require.Equal(t, "pending", replayed[0].State)
	require.NotEqual(t, originalID, replayed[0].ID) // fresh delivery_id, per R-19/R-20
}

func TestRunReplayExpansionCycleExcludesOriginalsOutsideWindow(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ctx := context.Background()
	tenantID, _ := testsupport.CreateTenant(t, pool)
	endpointID := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})
	testsupport.CreateDelivery(t, pool, tenantID, endpointID, testsupport.DeliveryOptions{
		State: "succeeded", CreatedAt: &replayBeforeWindow,
	})
	insertReplay(t, pool, endpointID, replayWindowStart, replayWindowEnd)

	_, err := worker.RunReplayExpansionCycle(ctx, pool)
	require.NoError(t, err)

	var count int
	require.NoError(t, pool.QueryRow(ctx, "SELECT count(*) FROM deliveries WHERE endpoint_id = $1 AND state != 'succeeded'", endpointID).Scan(&count))
	require.Equal(t, 0, count)
}

func TestRunReplayExpansionCycleExcludesPendingAndInFlightOriginals(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ctx := context.Background()
	tenantID, _ := testsupport.CreateTenant(t, pool)
	endpointID := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})
	testsupport.CreateDelivery(t, pool, tenantID, endpointID, testsupport.DeliveryOptions{State: "pending", CreatedAt: &replayWindowStart})
	testsupport.CreateDelivery(t, pool, tenantID, endpointID, testsupport.DeliveryOptions{State: "in_flight", CreatedAt: &replayWindowStart})
	insertReplay(t, pool, endpointID, replayWindowStart, replayWindowEnd)

	_, err := worker.RunReplayExpansionCycle(ctx, pool)
	require.NoError(t, err)

	var count int
	require.NoError(t, pool.QueryRow(ctx, "SELECT count(*) FROM deliveries WHERE endpoint_id = $1", endpointID).Scan(&count))
	require.Equal(t, 2, count) // only the two originals, no replayed rows added
}

func TestRunReplayExpansionCycleInsertsInOriginalChronologicalOrder(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ctx := context.Background()
	tenantID, _ := testsupport.CreateTenant(t, pool)
	endpointID := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})
	_, firstEventID := testsupport.CreateDelivery(t, pool, tenantID, endpointID, testsupport.DeliveryOptions{State: "succeeded", CreatedAt: &replayWindowStart})
	_, secondEventID := testsupport.CreateDelivery(t, pool, tenantID, endpointID, testsupport.DeliveryOptions{State: "failed", CreatedAt: &replayWindowStart})
	insertReplay(t, pool, endpointID, replayWindowStart, replayWindowEnd)

	_, err := worker.RunReplayExpansionCycle(ctx, pool)
	require.NoError(t, err)

	rows, err := pool.Query(ctx, "SELECT event_id FROM deliveries WHERE endpoint_id = $1 AND state = 'pending' ORDER BY seq", endpointID)
	require.NoError(t, err)
	var eventIDs []string
	for rows.Next() {
		var id string
		require.NoError(t, rows.Scan(&id))
		eventIDs = append(eventIDs, id)
	}
	require.Equal(t, []string{firstEventID, secondEventID}, eventIDs)
}

func TestRunReplayExpansionCycleReturnsFalseWhenNothingToExpand(t *testing.T) {
	pool := testsupport.SetupPool(t)
	did, err := worker.RunReplayExpansionCycle(context.Background(), pool)
	require.NoError(t, err)
	require.False(t, did)
}
